polaris/evaluate/_metric.py

```diff
@@ -84,55 +164,241 @@ class Metric(Enum):
     roc_auc_ovo = MetricInfo(
         fn=roc_auc_score, kwargs={"multi_class": "ovo"}, direction="max", y_type="y_score"
     )
-    # TODO: add metrics to handle multitask multiclass predictions.
 
     # docking related metrics
     rmsd_coverage = MetricInfo(fn=rmsd_coverage, direction="max", y_type="y_pred")
 
+
+class Metric(BaseModel):
+    """
+    A Metric in Polaris.
+
+    Any metric implementation in Polaris object wraps a BaseMetric. This class holds the most vanilla implementation.
+    It does some minor preprocessing on the inputs and then simply passes these arguments to the BaseMetric.
+
+    It serves as a base class for more complex metrics.
+
+    Attributes:
+        metric: The actual callable that is at the core of the metric implementation.
+        kind: The kind of metric, uses to discriminate different types of metrics in a discriminated union.
+        name: The name of the metric. The metrics for a specific benchmark need to be uniquely named.
+            In most cases, a sensible default will be set and the user doesn't need to worry about this.
+        config: For more complex metrics, this object should hold all parameters for the metric.
+    """
+
+    metric: BaseMetric
```

Comment:
This cleans things up a lot! Thanks! 

Two points: 
- Because the `MetricInfo` should not be touched by the user, I changed this class to be frozen and made it a property of `Metric`, rather than it being part of the data model.
- I wasn't sure why you suggested changing to a `TypedDict`. To me it seems a Pydantic model has support for more features at a negligible cost. What am I missing here?

Reply:
>     * Because the `MetricInfo` should not be touched by the user, I changed this class to be frozen and made it a property of `Metric`, rather than it being part of the data model.

The individual fields can also be set to frozen, which gives you the same safety. You avoid having to have a bunch of computed properties just to handle another level of indirection.

>     * I wasn't sure why you suggested changing to a `TypedDict`. To me it seems a Pydantic model has support for more features at a negligible cost. What am I missing here?

It wasn't clear to me, at that point of the review, what this was meant to contain. If the idea is to store just a grab bag of arbitrary data, then is simpler to use a dict. If there's behavior attached to that data, then I agree a model makes more sense.